presto-hive/src/main/java/com/facebook/presto/hive/security/ranger/RangerBasedAccessControl.java

```diff
@@ -0,0 +1,516 @@
+
+        JsonCodec<Users> usersJsonCodec = jsonCodec(Users.class);
+        JsonResponse<Users> users = JsonResponse.execute(usersJsonCodec, client, request);
+        if (!users.hasValue()) {
+            throw new RuntimeException(format("Request to %s failed: %s [Error: %s]", endPtUri, users, users.getResponseBody()));
+        }
+
+        return users.getValue();
+    }
+
+    @TestOnly
+    public void setRangerAuthorizer(RangerAuthorizer rangerAuthorizer)
+    {
+        this.rangerAuthorizer = rangerAuthorizer;
+    }
+
+    @TestOnly
+    public void setUsers(Users users)
+    {
+        this.users = users;
+    }
+
+    private static <T> T jsonParse(Response response, Class<T> clazz)
+            throws IOException
+    {
+        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(response.body().byteStream()));
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        return mapper.readValue(bufferedReader, clazz);
+    }
+
+    private boolean userExists(String userName)
+    {
+        return users.getvXUsers().stream()
+                .filter(user -> userName.equalsIgnoreCase(user.getName()))
+                .findAny().isPresent();
+    }
+
+    private Set<String> getGroupsForUser(String username)
+    {
+        if (userExists(username)) {
+            return users.getvXUsers().stream()
+                    .filter(user -> username.equalsIgnoreCase(user.getName()))
+                    .findFirst()
+                    .get()
+                    .getGroupNameList()
+                    .stream().collect(Collectors.toSet());
+        }
+        return null;
+    }
+
+    enum HiveAccessType
+    {
+        NONE, CREATE, ALTER, DROP, INDEX, LOCK, SELECT, UPDATE, USE, ALL, ADMIN
+    };
+
+    private boolean checkAccess(ConnectorIdentity identity, SchemaTableName tableName, String column, HiveAccessType accessType)
+    {
+        return rangerAuthorizer.authorizeHiveResource(tableName.getSchemaName(), tableName.getTableName(), column,
+                accessType.toString(), identity.getUser(), getGroupsForUser(identity.getUser()));
+    }
+
+    /**
+     * Check if identity is allowed to create the specified schema in this catalog.
+     *
+     * @throws AccessDeniedException if not allowed
+     */
+    public void checkCanCreateSchema(ConnectorTransactionHandle transactionHandle, ConnectorIdentity identity, AccessControlContext context, String schemaName)
+    {
+    }
+
+    /**
+     * Check if identity is allowed to drop the specified schema in this catalog.
+     *
+     * @throws AccessDeniedException if not allowed
+     */
+    public void checkCanDropSchema(ConnectorTransactionHandle transactionHandle, ConnectorIdentity identity, AccessControlContext context, String schemaName)
+    {
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, schemaName);
```

Comment:
Lets assume that the schema in question here is `s1` - I don't have previous experience with Ranger - what we want to check here is that if the user has permission to drop schema `s1`, but given this implementation - how will it differentiate between what we want and the permission to drop table `s1` in schema `s1`

Reply:
This implementation is keeping authorization checks similar to Hive since it is verifying against policies set for Hive service. 
Ranger authorization checks are validated and inherited from schema, table & column.
Ranger policy can define if accesses are across all tables in the schema or specific table. The policy can allow/restrict user or group for drop access across which will be leveraged for schema and then more specific & restrictive policies for a table can override that behavior. 
Policies are evaluated in a specific order to ensure predictable results.
More details here - https://cwiki.apache.org/confluence/display/RANGER/Deny-conditions+and+excludes+in+Ranger+policies
TestRangerBasedAccessControl has concrete tests around this.